sopel/lifecycle.py

```diff
@@ -0,0 +1,180 @@
+"""Deprecation module for Sopel developers and Plugin authors.
+
+.. versionadded:: 8.0
+
+    Previously in :mod:`sopel.tools`, the :func:`deprecated` function has been
+    moved to this newly created module, as it can be used in every part of the
+    Sopel codebase, including :mod:`sopel.tools` itself.
+
+"""
+from __future__ import annotations
+
+import functools
+import inspect
+import logging
+import traceback
+from typing import Callable, Optional
+
+from pkg_resources import parse_version
+
+from sopel import __version__
+
+
+def deprecated(
+    reason: Optional[str] = None,
+    version: Optional[str] = None,
+    removed_in: Optional[str] = None,
+    warning_in: Optional[str] = None,
+    stack_frame: int = -1,
+    func: Optional[Callable] = None,
+):
+    """Decorator to mark deprecated functions in Sopel's API
+
+    :param reason: optional text added to the deprecation warning
+    :param version: optional version number when the decorated function
+                        is deprecated
+    :param removed_in: optional version number when the deprecated function
+                           will be removed
+    :param warning_in: optional version number when the decorated function
+                           should start emitting a warning when called
+    :param stack_frame: optional stack frame to output; defaults to
+                            ``-1``; should almost always be negative
```

Comment:
```suggestion
    :param version: optional version number when the decorated function
                    is deprecated
    :param removed_in: optional version number when the deprecated function
                       will be removed
    :param warning_in: optional version number when the decorated function
                       should start emitting a warning when called
    :param stack_frame: optional stack frame to output; defaults to
                        ``-1``; should almost always be negative
```

I didn't look to see if this weird indentation was in the original `tools` definition, but… I'm gonna fix it blind.

Reply:
Yeah that's weird.